Anemoi hash r1cs numstatecolumns

vesselinux opened the pull request:
Rebase [anemoi-hash-r1cs-numstatecolumns](https://github.com/clearmatics/libsnark/tree/anemoi-hash-r1cs-numstatecolumns) onto base branch [anemoi-hash-r1cs](https://github.com/clearmatics/libsnark/tree/anemoi-hash-r1cs). See PR #79 [FROZEN] which was automatically closed after deleting branch [anemoi-hash-r1cs-parametrization](https://github.com/clearmatics/libsnark/tree/anemoi-hash-r1cs-parametrization) . This branch implements specialised implementations of multiplication by the MDS matrix for valid values of `NumStateColumns_L`. In contrast, the method used in the base branch is a single function for all matrix sizes (1,2,3,4) with static assert check on the dimensions (https://github.com/clearmatics/libsnark/blob/anemoi-hash-r1cs/libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc#L326). See the last bullet in https://github.com/clearmatics/libsnark/issues/77 and https://github.com/clearmatics/libsnark/pull/65#discussion_r992467709 .

Commit:
anemoi: removed duplicate gadgets for the flystel Q-functions for prime and binary fields (https://github.com/clearmatics/libsnark/issues/77, https://github.com/clearmatics/libsnark/pull/65#discussion_r992421012, https://github.com/clearmatics/libsnark/pull/65#discussion_r992423162) - resolved conflicts after rebase onto anemoi-hash-r1cs

dtebbs commented on libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp:
```diff
@@ -183,10 +183,16 @@ class flystel_prime_field_gadget : public gadget<libff::Fr<ppT>>
-std::array<std::array<FieldT, NumStateColumns_L>, NumStateColumns_L>
-anemoi_permutation_mds(const FieldT g);
+// get the MDS matrix for each allowed dimension: 2,3 or 4
+template<typename ppT>
+std::array<std::array<libff::Fr<ppT>, 2>, 2> anemoi_permutation_mds_2x2(
+    const libff::Fr<ppT> g);
+template<typename ppT>
+std::array<std::array<libff::Fr<ppT>, 3>, 3> anemoi_permutation_mds_3x3(
+    const libff::Fr<ppT> g);
+template<typename ppT>
+std::array<std::array<libff::Fr<ppT>, 4>, 4> anemoi_permutation_mds_4x4(
+    const libff::Fr<ppT> g);
```
These all have different names, so the code that uses them must still have `if-else` clauses based on the matrix size.  The idea behind the comment https://github.com/clearmatics/libsnark/pull/65#discussion_r992467709 was to use the size as a template parameter, so the calling code can just call `anemoi_permutation_mds<ppT, NumStateColumnsL>()`.

Note, as discussed w.r.t. `ppT`, `anemoi_parameters` etc, we can't template functions in precisely this way, so we template classes:
```
template<typename ppT, size_t NumStateColumns_L> class anemoi_permutation_mds;
```
where the class has a static of the form `permutation_mds()`, and we can give an implementation for `template<typename ppT> class anemoi_permutation_mds<ppT, 2>`, `template<typename ppT> class anemoi_permutation_mds<ppT, 3>`, `template<typename ppT> class anemoi_permutation_mds<ppT, 4>`.

Calling code can then use its `NumStateColumns_L` parameter and avoid any complex if statements by calling `anemoi_permutation_mds<ppT, NumStateColumnsL>::permutation_mds()`.  If calling code tries to use an invalid `NumStateColumns_L` param, it will get an error at compile (or, more precisely at link) time.  Note, we will probably also want to define the matrix types as part of the class.  This is the pattern that's used in many places in this code base, including `aemoi_parameters`.

vesselinux replied:
Understood. Fixed.

dtebbs replied:
>  Note, we will probably also want to define the matrix types as part of the class.

Not defined, but we can address that when this code is used.

vesselinux replied:
Do you mean having something like 

`using anemoi_mds_matrix_t = std::array<std::array<libff::Fr<ppT>, 2>, 2>;` 

inside each class?

vesselinux replied:
Added matrix type definitions.

Commit:
anemoi: implemented function for generating the mds matrix for different number of columns using class specialization (https://github.com/clearmatics/libsnark/pull/102#discussion_r1069518363)

dtebbs commented on libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp:
```diff
@@ -183,16 +183,29 @@ class flystel_prime_field_gadget : public gadget<libff::Fr<ppT>>
+{
+public:
+    static std::array<std::array<libff::Fr<ppT>, 2>, 2> permutation_mds(
+        const libff::Fr<ppT> g);
```
Do you ever use something other than `anemoi_parameters<ppT>::multiplicative_generator_g` here?  If not, can this just implicitly use that value?

vesselinux replied:
ATM I am only using `anemoi_parameters<ppT>::multiplicative_generator_g`. I really like the idea of using the latter implicitly instead of explicitly passing an extra input parameter `g` and I am tempted to apply the change (in fact, I already did but haven't pushed it yet). Note though that for some unit tests I am using the `debug_parameters_bls12_381` class which incidentally has the same value for `g=7`, but which can potentially be changed. So I wonder if I shouldn't keep the input parameter `g` after all, in order to accommodate a potential change of its value e.g. for unit testing. Happy to go either way, so interested to hear your opinion.

dtebbs replied:
Ah yes, that's a good point.  The aemoi_parameters is now a template parameter.  I'd forgotten that.  I guess we could make it a parameter of this class too, but it might get complex.  Ideally we'd extract the ppT part from the anemoi_parameters as well, so there were only 2 parameters.

Feel free to just put a note in the comments instead of making the change.  As you prefer.  This is all "internal" to the anemoi code anyway so it's easy to go back and change later without affecting too much other code.

vesselinux replied:
Ok, kept the input parameter `g` but added a TODO note as suggested.

Commit:
anemoi: added a TODO note regarding the removal of the input g parameter from all anemoi_permutation_mds::permutation_mds functions (https://github.com/clearmatics/libsnark/pull/102#discussion_r1071444422)

dtebbs commented:
Addresses #105

Commit:
anemoi: defined the types of the mds matrices (https://github.com/clearmatics/libsnark/pull/102#discussion_r1071214317)

Commit:
anemoi: fixed compilation errors related to method anemoi_permutation_mds::permutation_mds after rebase onto anemoi-hash-r1cs

vesselinux commented:
Rebased on `anemoi-hash-r1cs` and re-review requested. Please, pay special attention to: 

- `Note-1` (https://github.com/clearmatics/libsnark/blob/anemoi-hash-r1cs-numstatecolumns/libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc#L503-L507), 
- `Note-2` (https://github.com/clearmatics/libsnark/blob/anemoi-hash-r1cs-numstatecolumns/libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc#L509-L514) 
- `Note-3` (https://github.com/clearmatics/libsnark/blob/anemoi-hash-r1cs-numstatecolumns/libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc#L516-L524) 

added as comments in file `anemoi_components.tcc` .
